### ISSUE TITLE
Adopt iOS 17 `@Observable`?

### ISSUE THREAD

**Issue by defagos** [SRGSSR/pillarbox-apple#424]
### Detailed question

Apple [introduced](https://developer.apple.com/wwdc23/10149) `@Observable` at WWDC 2023. Should we adopt it for Pillarbox player?

### Can your answer be found elsewhere?

- [X] I have searched existing [issues](https://github.com/SRGSSR/pillarbox-apple/issues?q=is%3Aissue+label%3Aquestion+) and found no similar question.

- [X] I have browsed the available documentation and found no answer to my question.

**Comment by defagos** [comment]
It is probably a bad idea to adopt `@Observable` for `Player`, as a major difference between `@Observable` and `@ObservableObject` is that the latter is inherently reactive, which is exactly what is needed for Pillarbox player.

We could probably use `@Observable` for non-necessarily reactive components like `ProgressTracker` or `VisibilityTracker`, though, but only when Pillarbox officially supports iOS 17+.

**Comment by defagos** [comment]
[Pitch discussions](https://forums.swift.org/t/pitch-observation-revised/63757/30) are extremely interesting.

I remain convinced that we should keep `ObservableObject` which best suits our needs.

**Comment by defagos** [comment]
Once we have finished #559 we might be able to investigate `@Observable` support again. We will likely reach a state where no state is publicly published, with a publisher delivering updates instead. We should:

1. Hack `@Observable` support. According to @walid properties deeply nested in player properties can be automatically observed, leading to partial body updates. With hacked support we should be able to verify the UI behavior and the number of refreshes obtained pretty easily.
2. If the approach is successful we should transform our current `Player` into an implementation class `_Player` for iOS 16 `ObservableObject` support implemented by another `Player` class, simply forwarding methods to `_Player`. On iOS 17 this class would be marked as `@Observable`, probably with the help of the preprocessor. Two naive approaches:
    - Use the preprocesor to strip `@Observable` and associated attributes from `Player` on iOS 16, and to rename `Player` as `_Player`. Provide the separate `Player` implementation on iOS 16, conforming to `ObservableObject` and forwarding calls to the `_Player` implementation.
    - On iOS 17 use `Player` as is.

To provide this API officially, though, we should likely wait until [leaks in modals](https://developer.apple.com/forums/thread/736110?page=1) have been fixed.
